lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -699,6 +702,159 @@ void ProxyRenderDelegate::_UpdateSelectionStates()
+    //     been removed. We need this extra call to hide the previously shown
+    //     items.
+    //  3: Update the render tags with the new final render tags so that the next
+    //     real call to execute will update all the now visible items.
+    bool renderPurposeChanged, proxyPurposeChanged, guidePurposeChanged;
+    _proxyShapeData->UpdatePurpose(
+        &renderPurposeChanged, &proxyPurposeChanged, &guidePurposeChanged);
+    if (renderPurposeChanged || proxyPurposeChanged || guidePurposeChanged)
+    {
+        // Build the list of render tags which were added or removed (changed)
+        // and the list of render tags which were removed.
+        TfTokenVector changedRenderTags;
+        TfTokenVector removedRenderTags;
+        if (renderPurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->render);
+            if (!_proxyShapeData->DrawRenderPurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->render);
+        }
+        if (proxyPurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->proxy);
+            if (!_proxyShapeData->DrawProxyPurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->proxy);
+        }
+        if (guidePurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->guide);
+            if (!_proxyShapeData->DrawGuidePurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->guide);
+        }
+
+        // Mark all the rprims which have a render tag which changed dirty
+        SdfPathVector    rprimsToDirty = _GetFilteredRprims(*(_defaultCollection.get()), changedRenderTags);
+        HdChangeTracker& changeTracker = _renderIndex->GetChangeTracker();
+        for (auto id : rprimsToDirty) {
+            changeTracker.MarkRprimDirty(id, HdChangeTracker::DirtyRenderTag);
+        }
+
+        // Make a special pass over the removed render tags so that the objects
+        // can hide themselves.
+        _taskController->SetRenderTags(removedRenderTags);
+        _engine.Execute(_renderIndex.get(), &_dummyTasks);
+
+        // Set the new render tags correctly.The regular execute will cause
+        // any newly visible rprims will update and mark themselves visible.
+        TfTokenVector renderTags
+            = { HdRenderTagTokens->geometry }; // always draw geometry render tag purpose.
+        if (_proxyShapeData->DrawRenderPurpose()) {
+            renderTags.push_back(HdRenderTagTokens->render);
+        }
+        if (_proxyShapeData->DrawProxyPurpose()) {
+            renderTags.push_back(HdRenderTagTokens->proxy);
+        }
+        if (_proxyShapeData->DrawGuidePurpose()) {
+            renderTags.push_back(HdRenderTagTokens->guide);
+        }
+        _taskController->SetRenderTags(renderTags);
+    }
+}
+
+
+// Copied from renderIndex.cpp, the code that does HdRenderIndex::GetDrawItems. But I just want the rprimIds, I don't want to go all the way to draw items.
```

Comment:
These are all copied out of USD proper. I don't see a way to do it without executing some code like this. USD also has a few near-duplicate versions of this function. _DrawItemFilterPredicate and _DirtyListFilterPredicate are quite similar in how they filter for renderTags, _DirtyListFilterPredicate just adds the dirty filter as well.

Reply:
The better way of doing this would be to modify USD. HdRenderIndex::GetDrawItems() kind of does what I want, but it goes a step further in converting the rprimIds into HdDrawItems that I don't want. The functionality I want could get split out into a separate helper, essentially moving _GetFilteredRprims to HdRenderIndex and avoiding the code duplication.